### PR TITLE
[WEB-1007] FetchUserEmailQuery

### PR DESCRIPTION
**Pull request by scottkicks**
# 📲 What

In order to reduce the amount/size of network calls to get the currently logged in user's email for our CAPI events, we can add a simpler query that just gets the email and store it globally instead of asking for the entire user object every time.

# 🤔 Why

reduces the amount of network calls needed and makes accessing the user's email simpler.

# 🛠 How

Add a small one field query in AppDelegate that will grab the email and make it accessible app-wide in AppEnvironment.currency.loggedInUserEmail

### CODE REVIEW

File: KsApi/ServiceType.swift [hunk 1]
```diff
@@ -165,6 +165,10 @@ public protocol ServiceType {
   func fetchGraphUser(withStoredCards: Bool)
     -> SignalProducer<UserEnvelope<GraphUser>, ErrorEnvelope>
 
+  /// Fetches the email of the given User using graphQL.
```
Comment (msadoon) [inline KsApi/ServiceType.swift:168]: Mind saying that the user in this instance has to be logged in?

File: KsApi/models/UserEnvelope.swift [hunk 2]
```diff
@@ -15,4 +15,12 @@ extension UserEnvelope {
     }
     return SignalProducer(value: envelope)
   }
+
+  static func envelopeProducer(from data: GraphAPI.FetchUserEmailQuery.Data)
```
Comment (msadoon) [inline KsApi/models/UserEnvelope.swift:19]: nice work adding it here.

File: KsApi/models/graphql/GraphUserEmailTests.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      XCTAssertEqual(user.email, "blob@me.com")
+    } catch {
+      XCTFail("Failed to decode \(error)")
+    }
+  }
+
+  func testDecoding_WithNilValues() {
+    let jsonString = """
+    {
+      "email": "blob@me.com"
```
Comment (msadoon) [inline KsApi/models/graphql/GraphUserEmailTests.swift:31]: Should be `null`

**Review by msadoon — CHANGES_REQUESTED** [review]
Mostly there, just fix these issues.

**Comment by nativeksr** [comment]
### SwiftFormat found issues:

| File | Rules |
| ---- | ----- |
| Library/AppEnvironment.swift:45:1 | warning: (indent) Indent code in accordance with the scope level. |
| Library/Environment.swift:47:1 | warning: (indent) Indent code in accordance with the scope level. |
| Library/ViewModels/PledgePaymentMethodsViewModel.swift:388:1 | warning: (redundantParens) Remove redundant parentheses. |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by nativeksr** [comment]
### SwiftFormat found issues:

| File | Rules |
| ---- | ----- |
| Library/ViewModels/PledgeViewModelTests.swift:1965:1 | warning: (consecutiveBlankLines) Replace consecutive blank lines with a single blank line. |
| Library/ViewModels/PledgeViewModelTests.swift:2070:1 | warning: (consecutiveBlankLines) Replace consecutive blank lines with a single blank line. |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/ViewModels/PledgeViewModel.swift [hunk 4]
```diff
@@ -64,7 +64,7 @@ public protocol PledgeViewModelInputs {
   func riskMessagingViewControllerDismissed(isApplePay: Bool)
   func scaFlowCompleted(with result: StripePaymentHandlerActionStatusType, error: Error?)
   func shippingRuleSelected(_ shippingRule: ShippingRule)
-  func storeFacebookCAPIUserEmail(_ email: String?)
+
```
Comment (msadoon) [inline Library/ViewModels/PledgeViewModel.swift:67]: can probably remove this newline, format might do it, if not don't worry about it.

**Review by msadoon — CHANGES_REQUESTED** [review]
Some minor things here to update, but overall pretty good.
Mostly just get that signal firing from `userSessionStarted`.

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Kickstarter-iOS/AppDelegateViewModel.swift [hunk 5]
```diff
@@ -230,6 +230,23 @@ public final class AppDelegateViewModel: AppDelegateViewModelType, AppDelegateVi
           : SignalProducer(value: .value(nil))
       }
 
+    let fetchUserEmailEvent = Signal
+      .merge(
+        self.applicationWillEnterForegroundProperty.signal,
+        self.applicationLaunchOptionsProperty.signal.ignoreValues(),
```
[the remark this answers is not in the document]
Comment (scottkicks) [reply Kickstarter-iOS/AppDelegateViewModel.swift:236]: @msadoon I wanted to account for the case where users are already logged in when the feature is turned on. This way the next time they open the app we make sure to set their email. As opposed to waiting until they've logged out and then back in.

Does that make sense?

[the remark this answers is not in the document]
Comment (msadoon) [reply Kickstarter-iOS/AppDelegateViewModel.swift:236]: Yep, good call, didn't think about that.

**Review by msadoon — APPROVED** [review]
Good to go, just make that comment fix and tests pass.

**Pull request merged**